src/System.Private.CoreLib/shared/System/Threading/Semaphore.cs

```diff
@@ -0,0 +1,75 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.Win32;
+using Microsoft.Win32.SafeHandles;
+using System.IO;
+using System.Runtime.InteropServices;
+
+namespace System.Threading
+{
+    public sealed partial class Semaphore : WaitHandle
+    {
+        // creates a nameless semaphore object
+        // Win32 only takes maximum count of Int32.MaxValue
+        public Semaphore(int initialCount, int maximumCount) : this(initialCount, maximumCount, null) { }
+
+        public Semaphore(int initialCount, int maximumCount, string name)
+        {
+            VerifyCounts(initialCount, maximumCount);
+
+            bool createdNew;
+            CreateSemaphoreCore(initialCount, maximumCount, name, out createdNew);
+        }
+
+        public Semaphore(int initialCount, int maximumCount, string name, out bool createdNew)
+        {
+            VerifyCounts(initialCount, maximumCount);
```

Comment:
Nit: I'd suggest changing VerifyCounts to be named something more like ValidateArguments.  It might just be me, but when I see "VerifyCounts", I think it's doing something more with the internal state of the class.